# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1553113754 25200
#      Wed Mar 20 13:29:14 2019 -0700
# Node ID 7f86c097f29e709fddc2fdb2a4aefd6270f75ec5
# Parent  9e2f69c1a05d1fb2e0fe52c92840c32c3c73c555
methods.m: fix Java constructor post-processing with no package (bug #55858)

* methods.m: Fix regular expression to make package prefix optional.  Comment
all regular expressions to make the purpose clear.

diff --git a/scripts/miscellaneous/methods.m b/scripts/miscellaneous/methods.m
--- a/scripts/miscellaneous/methods.m
+++ b/scripts/miscellaneous/methods.m
@@ -71,20 +71,27 @@ function mtds = methods (obj, opt)
     mtds_list = strsplit (mtds_str, ';');
     mtds_list = mtds_list(:);  # return a column vector for compatibility
     havesigs = true;
   else
     error ("methods: invalid input argument");
   endif
 
   if (havesigs && ! showsigs)
+    ## Extract only the method name for ordinary class methods, delete the
+    ## return type and the argument list.
     mtds_list = regexprep (mtds_list, '^(?:[^(]+) (\w+) ?\(.*$', '$1');
-    mtds_list = regexprep (mtds_list, '^(?:[\.\w]+\.)(\w+) ?\(.*$', '$1');
+
+    ## Extract only the class name for class constructors, delete the optional
+    ## "org.example." package prefix and the argument list.
+    mtds_list = regexprep (mtds_list, '^(?:[\.\w]+\.)?(\w+) ?\(.*$', '$1');
+
     mtds_list = unique (mtds_list);
   else
+    ## Delete the "org.example." package prefix if present.
     mtds_list = regexprep (mtds_list, '^(?:[\.\w]+\.)(\w+ ?\(.*)$', '$1');
   endif
 
   if (nargout == 0)
     classname = ifelse (ischar (obj), obj, class (obj));
     printf ("Methods for class %s:\n", classname);
     disp (list_in_columns (mtds_list));
   else
